// @8base/sdk/lib/src/SchemaNameGenerator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var _ = require("lodash");
var pluralize = require("pluralize");
var makeFieldName = function (name) {
    return _.lowerFirst(name);
};
var makeTypeName = function (prefix, name) {
    return "" + _.upperFirst(prefix) + _.upperFirst(name);
};
var SchemaNameGenerator = {
    getTableItemFieldName: function (tableName) {
        return makeFieldName(pluralize.singular(tableName));
    },
    getTableListFieldName: function (tableName) {
        return makeFieldName(pluralize.plural(tableName) + "List");
    },
    getTableListResponseTypeName: function (tableName, typePrefix) {
        if (typePrefix === void 0) { typePrefix = ""; }
        return makeTypeName(typePrefix, pluralize.singular(tableName) + "ListResponse");
    },
    getCreateItemFieldName: function (tableName) {
        return makeFieldName(pluralize.singular(tableName) + "Create");
    },
    getUpdateItemFieldName: function (tableName) {
        return makeFieldName(pluralize.singular(tableName) + "Update");
    },
    getDeleteItemFieldName: function (tableName) {
        return makeFieldName(pluralize.singular(tableName) + "Delete");
    },
    getSmartTypeFieldName: function (fieldName) {
        return pluralize.singular(_.upperFirst(fieldName.toLowerCase()));
    },
    getOrderByInputTypeName: function (tableName, typePrefix) {
        if (typePrefix === void 0) { typePrefix = ""; }
        return makeTypeName(typePrefix, pluralize.singular(tableName) + "OrderBy");
    },
    getFilterInputTypeName: function (tableName, typePrefix) {
        if (typePrefix === void 0) { typePrefix = ""; }
        return makeTypeName(typePrefix, pluralize.singular(tableName) + "Filter");
    },
    getFieldsPermissionsInputTypeName: function (tableName, typePrefix) {
        if (typePrefix === void 0) { typePrefix = ""; }
        return makeTypeName(typePrefix, pluralize.singular(tableName) + "FieldsPermissions");
    },
    getEnumItemTypeName: function (tableName, fieldName, typePrefix) {
        if (typePrefix === void 0) { typePrefix = ""; }
        return makeTypeName(typePrefix, pluralize.singular(tableName) + "_" + _.upperFirst(fieldName) + "Enum");
    },
    getUpdateInputName: function (tableName, typePrefix) {
        if (typePrefix === void 0) { typePrefix = ""; }
        return makeTypeName(typePrefix, pluralize.singular(tableName) + "UpdateInput");
    },
    getDeleteInputName: function (tableName, typePrefix) {
        if (typePrefix === void 0) { typePrefix = ""; }
        return makeTypeName(typePrefix, pluralize.singular(tableName) + "DeleteInput");
    },
    getCreateInputName: function (tableName, typePrefix) {
        if (typePrefix === void 0) { typePrefix = ""; }
        return makeTypeName(typePrefix, pluralize.singular(tableName) + "CreateInput");
    },
    getRelationCreateInputName: function (tableName, relationFieldName, typePrefix) {
        if (typePrefix === void 0) { typePrefix = ""; }
        return makeTypeName(typePrefix, _.upperFirst(relationFieldName) + "_" + SchemaNameGenerator.getCreateInputName(tableName, typePrefix));
    },
    getRelationTableFilter: function (relationTableName, typePrefix) {
        if (typePrefix === void 0) { typePrefix = ""; }
        return makeTypeName(typePrefix, pluralize.singular(relationTableName) + "RelationFilter");
    },
    makeTableType: function (name, typePrefix) {
        if (typePrefix === void 0) { typePrefix = ""; }
        return makeTypeName(typePrefix, pluralize.singular(name));
    },
    getSubscriptionFilterName: function (tableName, typePrefix) {
        if (typePrefix === void 0) { typePrefix = ""; }
        return makeTypeName(typePrefix, pluralize.singular(tableName) + "SubscriptionFilter");
    },
    getSubscriptionPayload: function (tableName, typePrefix) {
        if (typePrefix === void 0) { typePrefix = ""; }
        return makeTypeName(typePrefix, pluralize.singular(tableName) + "Payload");
    },
    getSmartInputType: function (formatName) {
        return SchemaNameGenerator.getSmartTypeFieldName(formatName) + "Input";
    },
    getRelationName: function (tableName, relationFieldName, isUpdateField, typePrefix) {
        if (typePrefix === void 0) { typePrefix = ""; }
        var updateSuffix = (isUpdateField) ? 'Update' : '';
        return "" + makeTypeName(typePrefix, tableName) + _.upperFirst(relationFieldName) + updateSuffix + "RelationInput";
    },
};
exports.default = SchemaNameGenerator;
