// 8base/dist/interfaces/Extensions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ExtensionType;
(function (ExtensionType) {
    ExtensionType["resolver"] = "resolver";
    ExtensionType["trigger"] = "trigger";
    ExtensionType["webhook"] = "webhook";
    ExtensionType["task"] = "task";
})(ExtensionType = exports.ExtensionType || (exports.ExtensionType = {}));
/*
    Resolvers
*/
var GraphQLFunctionType;
(function (GraphQLFunctionType) {
    GraphQLFunctionType["Mutation"] = "Mutation";
    GraphQLFunctionType["Query"] = "Query";
})(GraphQLFunctionType = exports.GraphQLFunctionType || (exports.GraphQLFunctionType = {}));
/*
    Triggers
*/
var TriggerType;
(function (TriggerType) {
    TriggerType["before"] = "Before";
    TriggerType["after"] = "After";
})(TriggerType = exports.TriggerType || (exports.TriggerType = {}));
var TriggerOperation;
(function (TriggerOperation) {
    TriggerOperation["create"] = "Create";
    TriggerOperation["update"] = "Update";
    TriggerOperation["delete"] = "Delete";
})(TriggerOperation = exports.TriggerOperation || (exports.TriggerOperation = {}));
