// @8base/utils/dist/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SYSTEM_TABLES = exports.MUTATION_TYPE = exports.DATE_FORMATS = exports.SMART_FORMATS = exports.FILE_FORMATS = exports.NUMBER_FORMATS = exports.TEXT_FORMATS = exports.SWITCH_FORMATS = exports.FIELD_TYPE = void 0;
var FIELD_TYPE = {
  DATE: 'DATE',
  FILE: 'FILE',
  ID: 'ID',
  NUMBER: 'NUMBER',
  RELATION: 'RELATION',
  SWITCH: 'SWITCH',
  TEXT: 'TEXT',
  SMART: 'SMART'
};
exports.FIELD_TYPE = FIELD_TYPE;
var SWITCH_FORMATS = {
  ON_OFF: 'ON_OFF',
  YES_NO: 'YES_NO',
  TRUE_FALSE: 'TRUE_FALSE',
  ACTIVE_INACTIVE: 'ACTIVE_INACTIVE',
  HIGH_LOW: 'HIGH_LOW',
  SMART: 'SMART'
};
exports.SWITCH_FORMATS = SWITCH_FORMATS;
var TEXT_FORMATS = {
  UNFORMATTED: 'UNFORMATTED',
  NAME: 'NAME',
  ADDRESS: 'ADDRESS',
  SSN: 'SSN',
  EIN: 'EIN',
  PHONE: 'PHONE',
  EMAIL: 'EMAIL'
};
exports.TEXT_FORMATS = TEXT_FORMATS;
var NUMBER_FORMATS = {
  NUMBER: 'NUMBER',
  CURRENCY: 'CURRENCY',
  PERCENTAGE: 'PERCENTAGE',
  FRACTION: 'FRACTION',
  SCIENTIFIC: 'SCIENTIFIC'
};
exports.NUMBER_FORMATS = NUMBER_FORMATS;
var FILE_FORMATS = {
  FILE: 'FILE',
  IMAGE: 'IMAGE'
};
exports.FILE_FORMATS = FILE_FORMATS;
var SMART_FORMATS = {
  ADDRESS: 'ADDRESS',
  PHONE: 'PHONE'
};
exports.SMART_FORMATS = SMART_FORMATS;
var DATE_FORMATS = {
  DATE: 'DATE',
  DATETIME: 'DATE/TIME'
};
exports.DATE_FORMATS = DATE_FORMATS;
var MUTATION_TYPE = {
  CREATE: 'CREATE',
  UPDATE: 'UPDATE'
};
exports.MUTATION_TYPE = MUTATION_TYPE;
var SYSTEM_TABLES = {
  USERS: 'Users',
  FILES: 'Files',
  SETTINGS: 'Settings',
  ROLES: 'Roles',
  INVITATIONS: 'Invitations',
  PERMISSIONS: 'Permissions',
  API_TOKENS: 'ApiTokens',
  ENVIRONMENT_VARIABLES: 'EnvironmentVariables'
};
exports.SYSTEM_TABLES = SYSTEM_TABLES;
