// @8base/sdk/lib/src/subscriptions/types/SubscriptionInfo.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var graphql_tag_1 = require("graphql-tag");
var path = require("path");
var sha256 = require("sha256");
var SubscriptionInfo = (function () {
    function SubscriptionInfo(workspaceId, user) {
        this._filter = {};
        this.setUser(user);
        this.setWorkspaceId(workspaceId);
    }
    SubscriptionInfo.prototype.setUser = function (user) {
        this._user = user ? user : "";
        this._userhash = sha256(this._user);
    };
    Object.defineProperty(SubscriptionInfo.prototype, "user", {
        get: function () {
            return this._user;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SubscriptionInfo.prototype, "userHash", {
        get: function () {
            return this._userhash;
        },
        enumerable: true,
        configurable: true
    });
    SubscriptionInfo.prototype.setWorkspaceId = function (workspaceId) {
        this._workspaceId = workspaceId ? workspaceId : "";
    };
    Object.defineProperty(SubscriptionInfo.prototype, "workspaceId", {
        get: function () {
            return this._workspaceId;
        },
        enumerable: true,
        configurable: true
    });
    SubscriptionInfo.prototype.setFilter = function (filter) {
        this._filter = filter;
        return this;
    };
    Object.defineProperty(SubscriptionInfo.prototype, "filter", {
        get: function () {
            return this._filter;
        },
        enumerable: true,
        configurable: true
    });
    SubscriptionInfo.prototype.transformQuery = function (query) {
        return query.replace(/__typename/g, "");
    };
    SubscriptionInfo.prototype.getQueryHash = function (query) {
        return sha256(JSON.stringify((graphql_tag_1.default(templateObject_1 || (templateObject_1 = __makeTemplateObject(["", ""], ["", ""])), this.transformQuery(query))).definitions));
    };
    SubscriptionInfo.prototype.setQueryString = function (query) {
        if (query) {
            this._queryString = query;
            this._query = graphql_tag_1.default(templateObject_2 || (templateObject_2 = __makeTemplateObject(["", ""], ["", ""])), query);
            this._queryHash = this.getQueryHash(query);
        }
        return this;
    };
    Object.defineProperty(SubscriptionInfo.prototype, "query", {
        get: function () {
            return this._query;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SubscriptionInfo.prototype, "path", {
        get: function () {
            if (!this._queryHash) {
                throw new Error("Query is absent");
            }
            if (!this.action) {
                throw new Error("Action is absent");
            }
            return path.join(this.workspaceId, this._userhash, this.action, this._queryHash);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SubscriptionInfo.prototype, "header", {
        get: function () {
            return path.join(this.workspaceId, this._userhash);
        },
        enumerable: true,
        configurable: true
    });
    SubscriptionInfo.prototype.serialize = function () {
        return JSON.stringify({
            query: this._queryString,
            filter: this.filter,
            user: this.user,
            workspaceId: this.workspaceId
        });
    };
    SubscriptionInfo.prototype.deserialize = function (raw) {
        var parsed = JSON.parse(raw);
        this.setWorkspaceId(parsed.workspaceId);
        this.setUser(parsed.user);
        this.setQueryString(parsed.query);
        this.setFilter(parsed.filter);
        return this;
    };
    return SubscriptionInfo;
}());
exports.SubscriptionInfo = SubscriptionInfo;
var EntitySubscriptionInfo = (function (_super) {
    __extends(EntitySubscriptionInfo, _super);
    function EntitySubscriptionInfo(entity, workspaceId, user) {
        var _this = _super.call(this, workspaceId, user) || this;
        _this._entity = entity;
        return _this;
    }
    Object.defineProperty(EntitySubscriptionInfo.prototype, "action", {
        get: function () {
            return this._entity;
        },
        enumerable: true,
        configurable: true
    });
    return EntitySubscriptionInfo;
}(SubscriptionInfo));
exports.EntitySubscriptionInfo = EntitySubscriptionInfo;
var templateObject_1, templateObject_2;
