// expo-cli/build/commands/build/ios/appleApi/authenticate.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _slicedToArray2;

function _load_slicedToArray() {
  return _slicedToArray2 = _interopRequireDefault(require('babel-runtime/helpers/slicedToArray'));
}

var _regenerator;

function _load_regenerator() {
  return _regenerator = _interopRequireDefault(require('babel-runtime/regenerator'));
}

var _asyncToGenerator2;

function _load_asyncToGenerator() {
  return _asyncToGenerator2 = _interopRequireDefault(require('babel-runtime/helpers/asyncToGenerator'));
}

var _requestAppleIdCreds = function () {
  var _ref4 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee2(options) {
    return (_regenerator || _load_regenerator()).default.wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            _context2.t0 = _getAppleIdFromParams(options);

            if (_context2.t0) {
              _context2.next = 5;
              break;
            }

            _context2.next = 4;
            return _promptForAppleId();

          case 4:
            _context2.t0 = _context2.sent;

          case 5:
            return _context2.abrupt('return', _context2.t0);

          case 6:
          case 'end':
            return _context2.stop();
        }
      }
    }, _callee2, this);
  }));

  return function _requestAppleIdCreds(_x2) {
    return _ref4.apply(this, arguments);
  };
}();

var _promptForAppleId = function () {
  var _ref6 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee3() {
    var appleIdQuestions;
    return (_regenerator || _load_regenerator()).default.wrap(function _callee3$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            (0, (_log || _load_log()).default)('\nWe need your Apple ID/password to manage certificates, keys\nand provisioning profiles from your Apple Developer account.\n\n' + (_chalk || _load_chalk()).default.blue('Note: Expo does not keep your Apple ID or your Apple ID password.') + '\n  ');

            appleIdQuestions = [{
              type: 'input',
              name: 'appleId',
              message: 'What\'s your Apple ID?',
              validate: (_validators || _load_validators()).nonEmptyInput
            }, {
              type: 'password',
              name: 'appleIdPassword',
              message: 'Password?',
              validate: (_validators || _load_validators()).nonEmptyInput
            }];
            _context3.next = 4;
            return (0, (_prompt || _load_prompt()).default)(appleIdQuestions);

          case 4:
            return _context3.abrupt('return', _context3.sent);

          case 5:
          case 'end':
            return _context3.stop();
        }
      }
    }, _callee3, this);
  }));

  return function _promptForAppleId() {
    return _ref6.apply(this, arguments);
  };
}();

var _chooseTeam = function () {
  var _ref7 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee4(teams, userProvidedTeamId) {
    var foundTeam, _teams, team, choices, _ref9, _team;

    return (_regenerator || _load_regenerator()).default.wrap(function _callee4$(_context4) {
      while (1) {
        switch (_context4.prev = _context4.next) {
          case 0:
            if (!(teams.length === 0)) {
              _context4.next = 2;
              break;
            }

            throw new Error('You have no team associated with your Apple account, cannot proceed.\n(Do you have a paid Apple Developer account?)');

          case 2:
            if (!userProvidedTeamId) {
              _context4.next = 10;
              break;
            }

            foundTeam = teams.find(function (_ref8) {
              var teamId = _ref8.teamId;
              return teamId === userProvidedTeamId;
            });

            if (!foundTeam) {
              _context4.next = 9;
              break;
            }

            (0, (_log || _load_log()).default)('Using Apple Team with ID: ' + userProvidedTeamId);
            return _context4.abrupt('return', _formatTeam(foundTeam));

          case 9:
            (_log || _load_log()).default.warn('Your account is not associated with Apple Team with ID: ' + userProvidedTeamId);

          case 10:
            if (!(teams.length === 1)) {
              _context4.next = 16;
              break;
            }

            _teams = (0, (_slicedToArray2 || _load_slicedToArray()).default)(teams, 1), team = _teams[0];

            (0, (_log || _load_log()).default)('Only 1 team associated with your account, using Apple Team with ID: ' + team.teamId);
            return _context4.abrupt('return', _formatTeam(team));

          case 16:
            (0, (_log || _load_log()).default)('You have ' + teams.length + ' teams associated with your account');
            choices = teams.map(function (team, i) {
              return {
                name: i + 1 + ') ' + team.teamId + ' "' + team.name + '" (' + team.type + ')',
                value: team
              };
            });
            _context4.next = 20;
            return (0, (_prompt || _load_prompt()).default)({
              type: 'list',
              name: 'team',
              message: 'Which team would you like to use?',
              choices: choices
            });

          case 20:
            _ref9 = _context4.sent;
            _team = _ref9.team;
            return _context4.abrupt('return', _formatTeam(_team));

          case 23:
          case 'end':
            return _context4.stop();
        }
      }
    }, _callee4, this);
  }));

  return function _chooseTeam(_x3, _x4) {
    return _ref7.apply(this, arguments);
  };
}();

var _ora;

function _load_ora() {
  return _ora = _interopRequireDefault(require('ora'));
}

var _chalk;

function _load_chalk() {
  return _chalk = _interopRequireDefault(require('chalk'));
}

var _fastlane;

function _load_fastlane() {
  return _fastlane = require('./fastlane');
}

var _validators;

function _load_validators() {
  return _validators = require('../../../utils/validators');
}

var _log;

function _load_log() {
  return _log = _interopRequireDefault(require('../../../../log'));
}

var _prompt;

function _load_prompt() {
  return _prompt = _interopRequireDefault(require('../../../../prompt'));
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var APPLE_IN_HOUSE_TEAM_TYPE = 'in-house';

exports.default = function () {
  var _ref = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee(options) {
    var _ref2, appleId, appleIdPassword, spinner, _ref3, teams, team;

    return (_regenerator || _load_regenerator()).default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            _context.next = 2;
            return _requestAppleIdCreds(options);

          case 2:
            _ref2 = _context.sent;
            appleId = _ref2.appleId;
            appleIdPassword = _ref2.appleIdPassword;
            spinner = (0, (_ora || _load_ora()).default)('Trying to authenticate with Apple Developer Portal...').start();
            _context.prev = 6;
            _context.next = 9;
            return (0, (_fastlane || _load_fastlane()).runAction)((_fastlane || _load_fastlane()).travelingFastlane.authenticate, [appleId, appleIdPassword], {
              pipeStdout: true,
              onStdoutDataReceived: function onStdoutDataReceived() {
                return spinner.stop();
              }
            });

          case 9:
            _ref3 = _context.sent;
            teams = _ref3.teams;

            spinner.succeed('Authenticated with Apple Developer Portal successfully!');
            _context.next = 14;
            return _chooseTeam(teams, options.teamId);

          case 14:
            team = _context.sent;
            return _context.abrupt('return', { appleId: appleId, appleIdPassword: appleIdPassword, team: team });

          case 18:
            _context.prev = 18;
            _context.t0 = _context['catch'](6);

            spinner.fail('Authentication with Apple Developer Portal failed!');
            throw _context.t0;

          case 22:
          case 'end':
            return _context.stop();
        }
      }
    }, _callee, this, [[6, 18]]);
  }));

  function authenticate(_x) {
    return _ref.apply(this, arguments);
  }

  return authenticate;
}();

function _getAppleIdFromParams(_ref5) {
  var appleId = _ref5.appleId;

  var appleIdPassword = process.env.EXPO_APPLE_PASSWORD;
  if (appleId && appleIdPassword) {
    return {
      appleId: appleId,
      appleIdPassword: appleIdPassword
    };
  } else {
    return null;
  }
}

var _formatTeam = function _formatTeam(_ref10) {
  var teamId = _ref10.teamId,
      name = _ref10.name,
      type = _ref10.type;
  return {
    id: teamId,
    name: name + ' (' + type + ')',
    inHouse: type.toLowerCase() === APPLE_IN_HOUSE_TEAM_TYPE
  };
};
module.exports = exports['default'];
